src/lib/statistics/statistics_objects/equal_distinct_count_histogram.cpp

```diff
@@ -0,0 +1,223 @@
+  const BinID bin_count_with_extra_value = value_counts.size() % bin_count;
+
+  std::vector<T> bin_minima(bin_count);
+  std::vector<T> bin_maxima(bin_count);
+  std::vector<HistogramCountType> bin_heights(bin_count);
+
+  auto current_bin_begin_index = BinID{0};
+  for (BinID bin_index = 0; bin_index < bin_count; bin_index++) {
+    auto current_bin_end_index = current_bin_begin_index + distinct_count_per_bin - 1;
+    if (bin_index < bin_count_with_extra_value) {
+      current_bin_end_index++;
+    }
+
+    // We'd like to move strings, but have to copy if we need the same string for the bin_maximum
+    if (std::is_same_v<T, std::string> && current_bin_begin_index != current_bin_end_index) {
+      bin_minima[bin_index] = std::move(value_counts[current_bin_begin_index].first);
+    } else {
+      bin_minima[bin_index] = value_counts[current_bin_begin_index].first;
+    }
+
+    bin_maxima[bin_index] = std::move(value_counts[current_bin_end_index].first);
+
+    bin_heights[bin_index] =
+        std::accumulate(value_counts.cbegin() + current_bin_begin_index,
+                        value_counts.cbegin() + current_bin_end_index + 1, HistogramCountType{0},
+                        [](HistogramCountType a, const std::pair<T, HistogramCountType>& b) { return a + b.second; });
+
+    current_bin_begin_index = current_bin_end_index + 1;
+  }
+
+  return {std::move(bin_minima), std::move(bin_maxima), std::move(bin_heights),
+          static_cast<HistogramCountType>(distinct_count_per_bin), bin_count_with_extra_value};
+}
+
+template <typename T>
+std::shared_ptr<EqualDistinctCountHistogram<T>> EqualDistinctCountHistogram<T>::from_column(
+    const Table& table, const ColumnID column_id, const BinID max_bin_count, const HistogramDomain<T>& domain) {
+  Assert(max_bin_count > 0, "max_bin_count must be greater than zero ");
+
+  auto value_distribution = value_distribution_from_column(table, column_id, domain);
+
+  if (value_distribution.empty()) {
+    return nullptr;
+  }
+
+  auto bins = EqualDistinctCountHistogram<T>::_build_bins(std::move(value_distribution), max_bin_count);
+
+  return std::make_shared<EqualDistinctCountHistogram<T>>(std::move(bins.bin_minima), std::move(bins.bin_maxima),
+                                                          std::move(bins.bin_heights), bins.distinct_count_per_bin,
+                                                          bins.bin_count_with_extra_value, domain);
+}
+
+template <typename T>
+std::string EqualDistinctCountHistogram<T>::histogram_name() const {
+  return "EqualDistinctCount";
+}
+
+template <typename T>
+std::shared_ptr<AbstractHistogram<T>> EqualDistinctCountHistogram<T>::clone() const {
+  auto bin_minima = _bin_data.bin_minima;
+  auto bin_maxima = _bin_data.bin_maxima;
+  auto bin_heights = _bin_data.bin_heights;
+
+  return std::make_shared<EqualDistinctCountHistogram<T>>(std::move(bin_minima), std::move(bin_maxima),
```

Comment:
Since EqualDistinctCountHistogram::EqualDistinctCountHistogram() takes an rvalue ref, clang says no.
![image](https://user-images.githubusercontent.com/2679018/58906573-7f809000-870c-11e9-8891-e6acdd816744.png)